### DISCUSSION TITLE
What's the purpose of T in Handler?

### DISCUSSION THREAD

**Discussion by dominickpastore** [tokio-rs/axum#3024] [Q&A] [answered]
### Summary

`Handler` has this generic type parameter `T`. I'm looking at the docs and code, and it seems to be related to the extractors used by the handler, but I can't figure out why the trait needs it as one of its type parameters. The `call` function doesn't refer to `T` at all. The `Layered` and `HandlerService` structs do refer to `T`, but I can't figure out why they would need it either, except because the `Handler` they wrap needs it.

All in all, it seems like nothing would be lost if `Handler` (and `HandlerService` and `Layered`) didn't include `T` in their generic type parameters. I assume it's there for a reason though, so what am I missing here?

(The reason I ask is I'm trying to write a trait with a function that will return a `Handler`, but this `T` parameter is giving me a lot of trouble, and I don't understand its purpose well enough to figure out what to do about it.)

### axum version

0.7.7

**Comment by mladedav** [comment, marked as answer]
Feel free to play around with the code. You can try deleting it and see what happens. And honestly, I'd recommend doing just that because I'm not sure how much the following will make sense without you playing with the code.

The main reason is that we're creating multiple blanket implementations for the trait, one for `F: Fn()`, one for `F: Fn(T1)`, one for `F: Fn(T1, T2)`, etc. Normally, that wouldn't be possible because rustc would see that as conflicting implementations so we need to convince it it's actually slightly different traits. See [this example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=aaf1618561e7e4129d37365a061dcc8a), it might better show you what I'm trying to get across, the commented out code wouldn't work.

The second part of that parameter is usually named `M` which is the second parameter of `FromRequest` which exists for blanked implementation of `impl<T> FromRequestParts for T where T: FromRequest`, again solving issues with conflicting implementations.

**Reply by dominickpastore** [reply]
Thanks for the insight. I have a feeling I'm going to have some follow up questions about how to fix my trait later, but I'm going to try experimenting with the code like you suggested and see what I can learn from that first. Thanks!

**Comment by reivilibre** [comment]
Thanks a lot for asking this question; I was scratching my head trying to figure this out.

I'd like to add some text to the docstring explaining this (to save the next person), but not sure how exactly to phrase it. Does something like the following sound suitable?:

> The type parameter `T` is a workaround for trait coherence rules, allowing us to write blanket implementations
> of `Handler` over many types of handler functions with different numbers of arguments,
> without the compiler forbidding us from doing so because one type `F` can in theory
> implement both `Fn(T1)` and `Fn(T1, T2)`.
> `T` is a placeholder usually taking on a representation of the parameters of the handler function,
> allowing us to select one function signature to use as a `Handler`.
> See also: https://github.com/tokio-rs/axum/discussions/3024

I know the wording is a bit awkward (the concept is a bit awkward honestly) but I feel it at least is better than nothing.

**Reply by jplatte** [reply]
That makes sense to me 👍🏼
I wouldn't put the GitHub reference there, I don't think we link to GitHub elsewhere.
Also, note that the tuple we slot into `T` does not just contain all the argument types, it also contains the `M` generic of the final handler arguments' `FromRequest` impl (which _also_ exists on that trait to "defeat" coherence 😅).

**Reply by mladedav** [reply]
May I ask why you dug into the arguments? Were you trying to achieve a specific goal (like implementing a `Handler` for a specific usecase) or were you just curious?

I'm wondering who would be the intended audience of the documentation.

**Reply by reivilibre** [reply]
I am trying to write what is essentially a type-safe extension to `Router` to attach, in bulk, many `Handler`s in a foolproof and type-checked way, for what is essentially an RPC or API builder system.

Something like
```
router.handle_service(CatService {
    hiss: hiss_handler,
    meow: meow_handler,
    scratch: scratch_handler
})
```

Obviously `CatService` needs, at the very least, type params `<HHiss, HMeow, HScratch>`, but when mirroring the signature of `Router::post` a few times, I wondered what the `T` is and whether each Handler needs its own individual `T` or whether they would all be shared. (Answer now: my conclusion from this discussion is they all need their own `T` because they will have their own parameter lists)

Of course, it also feels as a matter of principle that any public parameter should have public documentation as to what it means, or at least how much you need to worry about it :-) — after all, you never know what crazy stuff people might do.
(I could also imagine feeling stuck if I was trying to decipher some compilation error that mentions `Handler<T, S>` and not having a clue what `T` was about. Maybe it actually wasn't important, but without any description of `T` I'd have no way of knowing that)

**Reply by reivilibre** [reply]
(The GitHub reference was mostly just in case people wanted to learn even more; hard to know how much to write / cut off — the playground link in https://github.com/tokio-rs/axum/discussions/3024#discussioncomment-11216721, I found extremely helpful, but not sure whether it is really appropriate to dive into so much detail here that it feels more like a Rust trait coherence tutorial :-))

**Reply by mladedav** [reply]
> whether each Handler needs its own individual T or whether they would all be shared. (Answer now: my conclusion from this discussion is they all need their own T because they will have their own parameter lists)

Just to make it clear, when you do custom implementations, you don't need to put the arguments in `T`. It exists solely so that you can implement the trait multiple times for the same type (which we don't do, we actually do it so that we can have multiple blanket implementations).

So if you create a `HissHandler` and want to implement `Handler` just once (I would not recommend trying to implement it more than once - type inference and ambiguity errors are sure to follow), you can put any type there including `()`, `Infallible`, or any custom `CatHandlerMarker`. And specifically, you can use the same `T` for all your handlers.

At the end of the day, the handler is called just with `Request` and your state, any other `FromRequestParts` parameters are constructed inside the `Handler` implementation itself.

**Reply by reivilibre** [reply]
I think we're on the same page about that, but I have a whole 'struct' of handlers (I want to ensure API implementors handle all the endpoints, rather than accidentally forget to attach some to the router).

In this case I need a struct with one `H` parameter for each handler, and I think one `T` parameter for each handler.


For a bit more of an illustrative example of why exactly I cared about the `T`s, here is the code I think I need to write:

```
pub struct CatService<HHiss, HMeow, HScratch> {
    hiss: HHiss,
    meow: HMeow,
    scratch: HScratch,
}

impl<HHiss, HMeow, HScratch> CatService<HHiss, HMeow, HScratch> {
    pub fn attach_to_router<S, THiss, TMeow, TScratch>(
        self,
        router: Router
    ) -> Router where
        HHiss: Handler<THiss, S>,
        HMeow: Handler<TMeow, S>,
        HScratch: Handler<TScratch, S>,
 {
        router.post("/hiss", self.hiss)
        .post("/meow", self.meow)
        .post("/scratch", self.scratch)
    }
}
```

maybe I'm wrong, but I don't entirely see how else it could be inferred if not done like this.

Edit:

And we need separate `T`s for all of these, because maybe the handler function passed to `hiss`
is going to be `async fn hiss_handler(auth: AuthExtractor, hiss: HissRequest) -> HissResponse` 
and maybe `meow_handler` doesn't have the exact same set of extractors
e.g. `async fn meow_handler(meow: MeowRequest) -> MeowResponse`

**Reply by mladedav** [reply]
Yeah, you're right, I misunderstood what you were trying to say.
